### PR TITLE
Upgrade deps, reduce bundle size

### PR DESCRIPTION
**Pull request by andykog**
* Updated to `mobx@4`
* Removed patched copy of `mobx-react` bundle. It is now bundled with remx from node_modules.
* Removed lodash as it increases bundle size dramatically (at least for those who don't already have this dependency)

### CODE REVIEW

**Comment by DanielZlotin** [comment]
Thanks for this. This PR is too big, please separate to 1 PR per bullet so we can tackle it faster

**Comment by DanielZlotin** [comment]
Also, the build failed. Separate PRs will also make the build easier to fix

**Comment by andykog** [comment]
Created #50 for API changes, don't want to split this PR anymore as there gonna be conflicts

File: src/es6Remx/connect.js [hunk 1]
```diff
@@ -1,11 +1,11 @@
-import * as _ from 'lodash';
+import isFunction from 'lodash.isfunction';
 import React from 'react';
-import { observer } from '../mobxReactClone'; // should import from mobx-react/custom when they fix issue #319
+import { observer } from 'mobx-react/custom'; // should import from mobx-react/custom when they fix issue #319
```
Comment (Niryo) [inline src/es6Remx/connect.js]: I am not sure we can remove the mobx-react clone. there was a problem with multiple mobx instances (see relevant issue), and I don't think they fixed it. maybe we can update the mobx-react version of the clone.

File: src/es6Remx/remx.js [hunk 2]
```diff
@@ -70,8 +70,8 @@ function isDev() {
 }
 
 export function toJS(data) {
-  console.warn(`remx.toJS() is deprecated. Please remove it from your code ASAP. 
-    Be aware that things can break after removing it, most of the time because of problematic data flow. 
+  console.warn(`remx.toJS() is deprecated. Please remove it from your code ASAP.
+    Be aware that things can break after removing it, most of the time because of problematic data flow.
     Please take your time to investigate the root of cause in case of a problem, toJS() is an expansive action`);
-  return cloneDeep(data);
+  return mobx.toJS(data);
```
Comment (Niryo) [inline src/es6Remx/remx.js]: is this call works when you give it something that is not mobx observable?

File: src/legacyRemx/connect.js [hunk 3]
```diff
@@ -1,7 +1,7 @@
-import { isFunction } from 'lodash';
+import isFunction from 'lodash.isfunction';
 import React from 'react';
 import { observable, action } from 'mobx';
-import { observer } from '../mobxReactClone'; // should import from mobx-react/custom when they fix issue #319
+import { observer } from 'mobx-react/custom'; // should import from mobx-react/custom when they fix issue #319
```
Comment (Niryo) [inline src/legacyRemx/connect.js]: same comment as above

File: package.json [hunk 4]
```diff
@@ -33,14 +33,17 @@
     "clean": "watchman watch-del-all && rm -rf dist"
   },
   "dependencies": {
+    "hoist-non-react-statics": "2.x.x",
     "lodash": "4.x.x",
-    "mobx": "3.x.x",
```
Comment (Niryo) [inline package.json]: why did you remove mobx from the dependencies?

File: src/es6Remx/connect.js (hunk 1 above)
Comment (andykog) [reply src/es6Remx/connect.js]: I've added `isolateGlobalState: true` in this PR, isn't that what you were looking for?

File: src/es6Remx/remx.js (hunk 2 above)
Comment (andykog) [reply src/es6Remx/remx.js]: `cloneDeep` comes from lodash, it doesn't know about observables, so, sure

File: package.json (hunk 4 above)
Comment (andykog) [reply package.json]: mobx has moved to devDependencies because it's being bundled into a single dist bundle by webpack (should help avoiding isolateGlobalState issues)

File: src/es6Remx/remx.js (hunk 2 above)
Comment (andykog) [reply src/es6Remx/remx.js]: Sorry, I mean, toJS does deep cloning for regular data structures, so it should work the same way as cloneDeep https://github.com/mobxjs/mobx/blob/master/src/api/tojs.ts

File: src/es6Remx/connect.js (hunk 1 above)
Comment (andykog) [reply src/es6Remx/connect.js]: Also, mobx & mobx-react are now bundled into a single file by webpack, so there’s no way that mobx-react will get a different mobx instance

**Comment by andykog** [comment]
@Niryo, I've decided to test this PR with WOA before merging and found that at least one module relies on mobx & mobx-react being brought by remx:
https://github.com/wix-private/tasks/blob/master/tasks-mobile/src/components/task/edit-task-store.ts#L4
https://github.com/wix-private/tasks/blob/master/tasks-mobile/src/components/task/fields/task-contact.component.tsx#L6
Sadly, but, I guess, for now we can't bundle mobx & mobx-react into a single file, otherwise those modules will have to bring these libs themselves, increasing WOA bundle size
